franka_lock_unlock/franka_shutdown.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python3
+# Copyright jk-ethz
+# Released under GNU AGPL-3.0
+# Contact us for other licensing options.
+
+# Developed and tested on system version
+# 4.2.1
+
+# Inspired by
+# https://github.com/frankaemika/libfranka/issues/63
+# https://github.com/ib101/DVK/blob/master/Code/DVK.py
+
+from itertools import count
+from time import sleep
+import argparse
+from requests.exceptions import ConnectionError
+from urllib.parse import urljoin
+from .franka_client import FrankaClient
+
+
+class FrankaShutdown(FrankaClient):
+    def __init__(self, hostname: str, username: str, password: str, protocol: str = 'https'):
+        super().__init__(hostname, username, password, protocol=protocol)
+
+    def _shutdown(self):
+        print("Shutting down...")
+        assert self._is_active_token(), "Cannot shutdown without an active control token."
+        try:
+            self._session.post(urljoin(self._hostname, '/admin/api/shutdown'), json={'token': self._token})
+        except ConnectionError as _:
+            # Sometimes, the server can shut down before sending a response, possibly raising an exception.
+            # Anyways, the server has still received the request, thus the robot shutdown procedure will start.
+            # So, we can ignore the cases when these exceptions are raised.
+            pass
+        finally:
+            print("The robot is shutting down. Please wait for the yellow lights to turn off, then switch the control box off.")
+
+    def run(self, wait: bool = False, request: bool = False) -> None:
+        assert not request or wait, "Requesting control without waiting for obtaining control is not supported."
+        self._login()
+        try:
+            assert self._token is not None or self._get_active_token_id() is None or wait, "Error requesting control, the robot is currently in use."
+            while True:
+                self._request_token(physically=False)
+                try:
+                    # Consider the timeout of 20 s for requesting physical access to the robot
+                    for _ in range(20) if request else count():
+                        if (not wait and not request) or self._is_active_token():
+                            print('Successfully acquired control over the robot.')
+                            return
+                        if request:
+                            print('Please press the button with the (blue) circle on the robot to confirm physical access.')
+                        elif wait:
+                            print('Please confirm the request message in the web interface on the logged in user.')
+                        sleep(1)
+                    # In case physical access was not confirmed, try again
+                    self._release_token()
+                finally:
+                    self._shutdown()
```

Comment:
I see there is a lot of code duplication from the `lock_unlock` file.
Two options:
- (Preferred) We could incorporate the `_shutdown` at the end of the `run` method in the `lock_unlock` script if an additional `--shutdown` arg is given to the script
- We could abstract away the common part of the code to the `FrankaClient`

With the current state here, the duplication makes it harder to maintain also different Franka versions in the future.

Reply:
Ok, not such a big blocker for me. But then I'd like to have the `_shutdown()` method in the `FrankaClient` at least. You can still have whatever flow you prefer separately in the `run`, but I don't see why we cannot move the `_shutdown` method into the client.